File: Conveyor_GitHub_Release_Final/window_manager.py
import window_utils
import win32gui
import win32con

class WindowManager:
    def __init__(self, screen_w, screen_h):
        self.screen_w = screen_w
        self.screen_h = screen_h
        self.managed_hwnds = []
        self.window_widths = {}
        self.window_gap = 5
        self.last_pos_cache = {}
        
        self.standard_width_factor = 0.5 # 50% screen width
        self.heavy_width_factor = 0.5    # Also 50% for consistency as requested
        self.floating_hwnds = set()
        self.last_active_hwnd = None

    def refresh_list(self, floating_classes=None, verbose=False):
        self.last_active_hwnd = win32gui.GetForegroundWindow()
        
        # Jika verbose, kita cetak header laporan
        if verbose: print("\n=== STARTING FULL WINDOW SCAN ===", flush=True)
        
        all_real = window_utils.get_managed_windows()
        
        # Filter based on user-toggled floating AND config-defined floating classes
        self.managed_hwnds = []
        for hwnd in all_real:
            cls = win32gui.GetClassName(hwnd)
            if hwnd in self.floating_hwnds: 
                if verbose: print(f"[Skipped] Title: '{win32gui.GetWindowText(hwnd)}', Reason: 'Manually Floating'", flush=True)
                continue
            if floating_classes and cls in floating_classes: 
                if verbose: print(f"[Skipped] Title: '{win32gui.GetWindowText(hwnd)}', Reason: 'Config Floating Class'", flush=True)
                continue
            self.managed_hwnds.append(hwnd)

        if verbose: print(f"=== SCAN FINISHED. Managed: {len(self.managed_hwnds)} windows ===\n", flush=True)

        for hwnd in self.managed_hwnds:
            if hwnd not in self.window_widths:
                factor = self.heavy_width_factor if window_utils.is_heavy_app(hwnd) else self.standard_width_factor
                self.window_widths[hwnd] = int(self.screen_w * factor)

    def toggle_floating(self, hwnd):
        if not hwnd: return False
        if hwnd in self.floating_hwnds:
            self.floating_hwnds.remove(hwnd)
            # Set back to Not Topmost when returning to tiling strip
            win32gui.SetWindowPos(hwnd, win32con.HWND_NOTOPMOST, 0, 0, 0, 0, 
                                 win32con.SWP_NOMOVE | win32con.SWP_NOSIZE | win32con.SWP_NOACTIVATE)
        else:
            self.floating_hwnds.add(hwnd)
            # Set to Topmost when floating
            win32gui.SetWindowPos(hwnd, win32con.HWND_TOPMOST, 0, 0, 0, 0, 
                                 win32con.SWP_NOMOVE | win32con.SWP_NOSIZE | win32con.SWP_NOACTIVATE)
            if hwnd in self.managed_hwnds:
                self.managed_hwnds.remove(hwnd)
        self.last_pos_cache.clear()
        return True

    def sync_states(self, floating_classes=None):
        """Purely syncs the list. Removal of windows naturally causes the layout to tighten."""
        all_real = window_utils.get_managed_windows()
        
        current_list = []
        for hwnd in all_real:
            cls = win32gui.GetClassName(hwnd)
            
            # Jika ini Game atau Aplikasi Berat, paksa masuk ke barisan tiling
            if window_utils.is_heavy_app(hwnd):
                if hwnd in self.floating_hwnds:
                    self.floating_hwnds.remove(hwnd)
                    # Pastikan status topmost dicabut agar bisa ikut digeser
                    win32gui.SetWindowPos(hwnd, win32con.HWND_NOTOPMOST, 0, 0, 0, 0, 
                                         win32con.SWP_NOMOVE | win32con.SWP_NOSIZE | win32con.SWP_NOACTIVATE)
                current_list.append(hwnd)
                continue
                
            if hwnd in self.floating_hwnds: continue
            if floating_classes and any(fc.lower() in cls.lower() for fc in floating_classes): continue
            current_list.append(hwnd)
        
        focused_hwnd = win32gui.GetForegroundWindow()
        
        if current_list == self.managed_hwnds:
            if focused_hwnd in self.managed_hwnds:
                self.last_active_hwnd = focused_hwnd
            return False

        new_windows = [h for h in current_list if h not in self.managed_hwnds]
        removed_windows = [h for h in self.managed_hwnds if h not in current_list]

        # 1. REMOVE
        for h in removed_windows:
            if h in self.managed_hwnds:
                self.managed_hwnds.remove(h)
                if h in self.last_pos_cache: del self.last_pos_cache[h]

        # 2. ADD: Smart Insertion (to the right of last active)
        if new_windows:
            try:
                target_hwnd = self.last_active_hwnd
                if target_hwnd in self.managed_hwnds:
                    idx = self.managed_hwnds.index(target_hwnd)
                    for i, h in enumerate(new_windows):
                        self.managed_hwnds.insert(idx + 1 + i, h)
                else:
                    self.managed_hwnds.extend(new_windows)
            except:
                self.managed_hwnds.extend(new_windows)

        if focused_hwnd in self.managed_hwnds:
            self.last_active_hwnd = focused_hwnd

        for h in new_windows:
            if h not in self.window_widths:
                factor = self.heavy_width_factor if window_utils.is_heavy_app(h) else self.standard_width_factor
                self.window_widths[h] = int(self.screen_w * factor)
                
        # Force cache clear to ensure immediate 'tightening' of the strip
        if removed_windows or new_windows:
            self.last_pos_cache.clear()
            return True
            
        return False

    def get_width(self, hwnd):
        return self.window_widths.get(hwnd, int(self.screen_w * self.standard_width_factor))

    def calculate_layout(self, offset_x):
        """Tightens the layout by iterating through the current list sequentially.
        Optimized with view clipping for performance with 15+ windows.
        """
        layout = []
        current_x = int(-offset_x)
        
        # Buffer of 1 screen width to ensure smooth entry/exit
        buffer = self.screen_w
        
        for hwnd in self.managed_hwnds:
            w = int(self.get_width(hwnd))
            
            # View Clipping: Only update if window is within or near the viewport
            if (current_x + w > -buffer) and (current_x < self.screen_w + buffer):
                target_rect = (current_x, 0, w, self.screen_h)
                if self.last_pos_cache.get(hwnd) != target_rect:
                    layout.append((hwnd, current_x, 0, w, self.screen_h, False))
                    self.last_pos_cache[hwnd] = target_rect
            
            # The next window starts exactly after the current one + gap
            current_x += w + self.window_gap
            
        return layout

    def swap_windows(self, idx1, idx2):
        if 0 <= idx1 < len(self.managed_hwnds) and 0 <= idx2 < len(self.managed_hwnds):
            self.managed_hwnds[idx1], self.managed_hwnds[idx2] = self.managed_hwnds[idx2], self.managed_hwnds[idx1]
            self.last_pos_cache.clear()
            return True
        return False

File: Conveyor_CPP_version/src/WindowManager.cpp
#include "../include/WindowManager.h"

WindowManager::WindowManager() {}

void WindowManager::RefreshList(const Config& config) {
    lastActiveHwnd = GetForegroundWindow();
    std::vector<HWND> allReal;
    
    auto enumProc = [](HWND hwnd, LPARAM lParam) -> BOOL {
        std::vector<HWND>* list = (std::vector<HWND>*)lParam;
        list->push_back(hwnd);
        return TRUE;
    };
    
    std::vector<HWND> temp;
    EnumWindows(enumProc, (LPARAM)&temp);
    
    managedHwnds.clear();
    for (HWND hwnd : temp) {
        if (Utils::IsRealWindow(hwnd, config.floatingClasses)) {
            bool isFloating = false;
            for (HWND fh : floatingHwnds) if (fh == hwnd) { isFloating = true; break; }
            if (!isFloating) managedHwnds.push_back(hwnd);
        }
    }
    
    RECT wa = Utils::GetWorkArea();
    int sw = wa.right - wa.left;
    
    for (HWND hwnd : managedHwnds) {
        if (windowWidths.find(hwnd) == windowWidths.end()) {
            double factor = Utils::IsHeavyApp(hwnd) ? config.heavyWidthFactor : config.standardWidthFactor;
            windowWidths[hwnd] = (int)(sw * factor);
        }
    }
}

bool WindowManager::SyncStates(const Config& config) {
    std::vector<HWND> temp;
    auto enumProc = [](HWND hwnd, LPARAM lParam) -> BOOL {
        std::vector<HWND>* list = (std::vector<HWND>*)lParam;
        list->push_back(hwnd);
        return TRUE;
    };
    EnumWindows(enumProc, (LPARAM)&temp);
    
    std::vector<HWND> currentList;
    for (HWND hwnd : temp) {
        if (Utils::IsRealWindow(hwnd, config.floatingClasses)) {
            bool isFloating = false;
            for (HWND fh : floatingHwnds) if (fh == hwnd) { isFloating = true; break; }
            if (!isFloating) currentList.push_back(hwnd);
        }
    }
    
    HWND focused = GetForegroundWindow();
    if (currentList == managedHwnds) {
        bool inManaged = false;
        for (HWND h : managedHwnds) if (h == focused) { inManaged = true; break; }
        if (inManaged) lastActiveHwnd = focused;
        return false;
    }
    
    std::vector<HWND> newWindows;
    for (HWND h : currentList) {
        bool exists = false;
        for (HWND mh : managedHwnds) if (mh == h) { exists = true; break; }
        if (!exists) newWindows.push_back(h);
    }
    
    std::vector<HWND> removedWindows;
    for (HWND h : managedHwnds) {
        bool exists = false;
        for (HWND cl : currentList) if (cl == h) { exists = true; break; }
        if (!exists) removedWindows.push_back(h);
    }
    
    // Remove
    for (HWND h : removedWindows) {
        managedHwnds.erase(std::remove(managedHwnds.begin(), managedHwnds.end(), h), managedHwnds.end());
        lastPosCache.erase(h);
    }
    
    // Add
    if (!newWindows.empty()) {
        auto it = std::find(managedHwnds.begin(), managedHwnds.end(), lastActiveHwnd);
        if (it != managedHwnds.end()) {
            managedHwnds.insert(it + 1, newWindows.begin(), newWindows.end());
        } else {
            managedHwnds.insert(managedHwnds.end(), newWindows.begin(), newWindows.end());
        }
    }
    
    if (std::find(managedHwnds.begin(), managedHwnds.end(), focused) != managedHwnds.end()) {
        lastActiveHwnd = focused;
    }
    
    RECT wa = Utils::GetWorkArea();
    int sw = wa.right - wa.left;
    for (HWND h : newWindows) {
        if (windowWidths.find(h) == windowWidths.end()) {
            double factor = Utils::IsHeavyApp(h) ? config.heavyWidthFactor : config.standardWidthFactor;
            windowWidths[h] = (int)(sw * factor);
        }
    }
    
    if (!removedWindows.empty() || !newWindows.empty()) {
        lastPosCache.clear();
        return true;
    }
    return false;
}

void WindowManager::UpdateLayout(int screenWidth, int screenHeight, int workAreaY, int workAreaHeight, double offset_x, int gap) {
    int current_x = (int)(-offset_x);
    int buffer = screenWidth;
    
    for (HWND hwnd : managedHwnds) {
        int w = windowWidths[hwnd];
        if ((current_x + w > -buffer) && (current_x < screenWidth + buffer)) {
            RECT target = {current_x, workAreaY, current_x + w, workAreaY + workAreaHeight};
            RECT last = lastPosCache[hwnd];
            if (last.left != target.left || last.top != target.top || (last.right - last.left) != w || (last.bottom - last.top) != workAreaHeight) {
                Utils::SetWindowPosOptimized(hwnd, current_x, workAreaY, w, workAreaHeight);
                lastPosCache[hwnd] = target;
            }
        }
        current_x += w + gap;
    }
}

void WindowManager::ToggleFloating(HWND hwnd) {
    if (!hwnd) return;
    auto it = std::find(floatingHwnds.begin(), floatingHwnds.end(), hwnd);
    if (it != floatingHwnds.end()) {
        floatingHwnds.erase(it);
        SetWindowPos(hwnd, HWND_NOTOPMOST, 0, 0, 0, 0, SWP_NOMOVE | SWP_NOSIZE | SWP_NOACTIVATE);
    } else {
        floatingHwnds.push_back(hwnd);
        SetWindowPos(hwnd, HWND_TOPMOST, 0, 0, 0, 0, SWP_NOMOVE | SWP_NOSIZE | SWP_NOACTIVATE);
        managedHwnds.erase(std::remove(managedHwnds.begin(), managedHwnds.end(), hwnd), managedHwnds.end());
    }
    lastPosCache.clear();
}

bool WindowManager::SwapWindows(int idx1, int idx2) {
    if (idx1 >= 0 && idx1 < managedHwnds.size() && idx2 >= 0 && idx2 < managedHwnds.size()) {
        std::swap(managedHwnds[idx1], managedHwnds[idx2]);
        lastPosCache.clear();
        return true;
    }
    return false;
}
